### PR TITLE
Signup Page Input Form Validation

### PR DESCRIPTION
**Pull request by DayRB25**
A series of functions to better validate input data from sign up form (previously only checking for non-empty fields)
1. Validated email pattern
"something@of.thisform"
3. Validate that password contains at least 8 characters, one uppercase character, and one special character
4. Validated student's year consists of only numbers and is greater than 0
5. Validated that city field does not contain any numbers
6. Validated that state field contains no numbers and is longer than two characters (to prevent someone from entering state's abbreviation by mistake)
7. Handled error codes from server with more specificity
400 = "Username or email taken"
500 = "Something went wrong..."

Alert messages display on screen to inform user of relevant issue. 

Examples:
Invalid email:
image

Invalid Password Message:
image

Invalid year:
image

Invalid city:
image

Invalid state:
image

Username or email taken:
image

### CODE REVIEW

File: advisor-ui/src/pages/Signup/Signup.jsx [hunk 1]
```diff
@@ -60,8 +60,131 @@ export default function Signup() {
       locationState === "" ||
       address === ""
     ) {
+      return false;
+    }
+    return true;
+  };
+
+  const validateEmail = () => {
```
Comment (Brikky) [inline advisor-ui/src/pages/Signup/Signup.jsx:68]: You could probably accomplish all of this with a regex, something like `/^\S+@\S+\.\S+$/`.

Regexes to actually validate emails against the spec are pretty beefy but probably not worth worrying about. 

HTML forms/inputs might also have email as a type nowadays.

File: advisor-ui/src/pages/Signup/Signup.jsx [hunk 2]
```diff
@@ -60,8 +60,131 @@ export default function Signup() {
...
+        password.charCodeAt(i) <= upperCaseZ_Ascii
+      ) {
+        hasUpperCase = true;
+      }
+      if (specialCharacters.has(password[i])) {
+        hasSpecialCharacter = true;
+      }
+    }
+
+    return hasSpecialCharacter && hasUpperCase;
```
Comment (Brikky) [inline advisor-ui/src/pages/Signup/Signup.jsx:124]: Couple small comments:

- You're not checking that the PW has a lowercase character, IDK if that's what you were going for but I think that's a pretty common requirement
- You're checking for a specific set of special characters, but not actually restricting any characters. I think generally the intent behind restricting characters is to eliminate/reduce risk of things like SQL injection or passing invalid characters/unsupported characters.

**Review by Brikky — COMMENTED** [review]
The validations and stuff are common and definitely something I'd expect of an actual product that is user-facing/deployed, but IDK if it's really worth prioritizing right now.

I presume you'll be driving any demos of your product, so I'd recommend focusing on user stories that are your app's core functionality, and handling these refinements later/maybe add a `TODO`.

**Review by ghost — COMMENTED** [review]
Good to see that you are thinking about validating input from an untrusted source.

Keep in mind that the email(and other fields) should be validated server side as well. Some initial checks here to prompt user to fix input are good but more complete checks could be left for the backend.

**Pull request merged**